### PR TITLE
Restrict build info loading to ES jar, not any jar

### PR DESCRIPTION
**Pull request by rjernst**
This change makes the build info initialization only try to load a jar
manifest if it is the elasticsearch jar. Anything else (eg a repackaged
ES for use of transport client in an uber jar) will contain "Unknown"
for the build info as it does for tests currently.

fixes #21955

### CODE REVIEW

**Review by jasontedor — APPROVED** [review]
The change looks good to me and I think we should get it in as soon as we can, but I would like to manually verify that this addresses the issues for deploying to an application container before this integrated and I opened #24050 so that having to do this manually is a one-time thing.

**Comment by jasontedor** [comment]
So manual testing shows that we'll need a little more here. We can discuss tomorrow.

File: core/src/main/java/org/elasticsearch/Build.java [hunk 1]
```diff
@@ -43,8 +43,10 @@
         final String date;
         final boolean isSnapshot;
 
+        final String esPrefix = "elasticsearch-" + Version.CURRENT;
         final URL url = getElasticsearchCodebase();
-        if (url.toString().endsWith(".jar")) {
+        final String urlStr = url.toString();
+        if (urlStr.startsWith("file://") && (urlStr.endsWith(esPrefix + ".jar") || urlStr.endsWith(esPrefix + "-SNAPSHOT.jar"))) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Build.java]: This isn't quite right. Wrap the ends with checks in parentheses.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Just one thing.

File: core/src/main/java/org/elasticsearch/Build.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Build.java]: So it's starts with and (ends with or ends with).

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Build.java]: Also I think it is one slash only, annoyingly.

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/Build.java]: Not sure what you mean.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/Build.java]: Oh, I misread it (mobile phone, sorry). The only thing that needs to change then is file:// -> file:/.

**Comment by jasontedor** [comment]
I've tested this change by deploying a war that uses the `TransportClient` to Wildfly 10 and it's good.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**